src/JustEat.StatsD/V2/Buffer.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Buffers.Text;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace JustEat.StatsD.V2
+{
+    internal ref struct Buffer
+    {
+        public Buffer(Span<byte> source) : this()
+        {
+            Tail = source;
+            Written = 0;
+        }
+
+        public Span<byte> Tail;
+        public int Written;
+    }
+
+    internal static class BufferImpl
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TryWriteBytes(this ref Buffer src, Span<byte> bytes)
+        {
+            if (bytes.Length > src.Tail.Length) return false;
+
+            bytes.CopyTo(src.Tail);
+            src.Tail = src.Tail.Slice(bytes.Length);
+            src.Written += bytes.Length;
+            return true;
+        }
+
+        public static bool TryWriteUtf8String(this ref Buffer src, string str)
+        {
+#if NETCOREAPP2_1
+            try
+            {
+                var written = Encoding.UTF8.GetBytes(str, src.Tail);
```

Comment:
Would storing `Encoding.UTF8` in a static field improve the perf slightly by removing the need to keep accessing it through the getter?